File: python3/binary_tree_paths.py
from .model import TreeNode

# Space : O(n) 
# Time  : O(n)

class Solution:
    def binaryTreePaths(self, root: TreeNode) -> List[str]:
        if not root:
            return []
        
        ans = []
        stack = [(root, "{}".format(root.val))]
        
        while stack:
            temp = []
            for node, path in stack:
                if node.left:
                    temp.append((node.left, path + "->{}".format(node.left.val)))
                if node.right:
                    temp.append((node.right, path + "->{}".format(node.right.val)))
                if not node.left and not node.right:
                    ans.append(path)
        
            stack = temp
        
        return ans

File: golang/binary_tree_paths.go
package main

import (
	"fmt"
	"golang/model"
)

// Space : O(n)
// Time  : O(n)

type NodePathPair struct {
	N *model.TreeNode
	P string
}

func binaryTreePaths(root *model.TreeNode) []string {
	var ans []string
	if root == nil {
		return ans
	}

	var stack = []NodePathPair{
		{root, fmt.Sprintf("%d", root.Val)},
	}

	for len(stack) > 0 {
		var temp []NodePathPair
		for _, pair := range stack {
			if pair.N.Left != nil {
				temp = append(temp, NodePathPair{pair.N.Left, pair.P + fmt.Sprintf("->%d", pair.N.Left.Val)})
			}
			if pair.N.Right != nil {
				temp = append(temp, NodePathPair{pair.N.Right, pair.P + fmt.Sprintf("->%d", pair.N.Right.Val)})
			}
			if pair.N.Left == nil && pair.N.Right == nil {
				ans = append(ans, pair.P)
			}
		}
		stack = temp
	}

	return ans
}
